impl/src/main/java/org/jboss/weld/context/beanstore/AttributeBeanStore.java

```diff
@@ -160,7 +188,21 @@ protected NamingScheme getNamingScheme() {
     }
 
     public Iterator<BeanIdentifier> iterator() {
-        return beanStore.iterator();
+        Iterator<BeanIdentifier> iterator;
```

Comment:
When attaching the store we always copy local entries to the backing store. In addition, writes always go directly to the backing store if attached. Therefore, I think we could simplify this logic:

1. If (!isAttached() || !isAttributeLazyFetchingEnabled()) then return beanStore.iterator();
2. else (meaning both attached and lazy) then it is enough to return getPrefixedAttributeNames() iterator mapped by getNamingScheme().deprefix() function (because all the entries are in the backing store anyway). This may either be implemented as an on-demand transforming iterator (using our Iterators/Iterables utils) or by eagerly collecting the results in a collection (as we do now). The former approach may potentionally be a little more performant perhaps.